### DISCUSSION TITLE
Expand the [NotNull...] attributes to cover collections

### DISCUSSION THREAD

**Discussion by SRNissen** [dotnet/csharplang#7829] [Language Ideas]
Consider a method `static boool AnyNull(...)` used like:

```csharp
    object? NullForgivingMethod(object? a, object? b object? c)
    {
        if(AnyNull(a,b,c) return null; // <-- here
        else              return NullEnforcingMethod(a,b,c);
    }
```

Here, if `AnyNull` has overloads for 1 through X parameters, each parameter can be annotated with `[NotNullAttribute]`.

But if `AnyNull` is written like `(params object?[] objects)` there is no way to indicate that each object has been checked for null.

I would like to be able to do that.

**Comment by TahirAhmadov** [comment]
Unfortunately, this is probably almost impossible. How does C# know that you checked all the elements? It now has to interpret a loop's conditions - like `for(int i = 0; i < objects.Length; ++i) { ... }` - then `foreach(var x in objects) { ... }` - then how to interpret what's inside the `{ ... }`?

**Reply by SRNissen** [reply]
It doesn't, I would have to tell it that I have - like it already works with `[NotNullWhenNotNull(param)]` or `[NotNullWhen(expression)]`.

**Reply by TahirAhmadov** [reply]
```cs
bool AnyNull([ElementsNotNullWhen(true)] params object?[] objects)
{
  for(int i = 0; i < objects.Length; ++i)
  {
    if(objects[i] == null) return false;
  }
  return true;
}
```
I'm talking about the implementation of `AnyNull` method. In order to promise to the outside world that `objects` are all not-null, inside of the `AnyNull` the analyzer would need to interpret the loop logic to ensure that you actually checked each element for not-null, and that's the difficult (impossible) part.

**Reply by SRNissen** [reply]
> the analyzer would need to interpret the loop logic to ensure that you actually checked each element 

No.

If the analyzer could do that, we wouldn't need these annotations in the first place.

The point of annotations like this are that I, the programmer writing `AnyNull`, know things the analyzer cannot find out, and so I tell the analyzer "If this function returns false, you can consider the annotated parameters not null for analysis purposes."

Contrast and compare these two toy examples that already compile in existing C#:

**Without annotations**

```csharp
    public static bool AnyNull(object? obj1,
                               object? obj2,
                               object? obj3)
    {
        return obj1 == null
            || obj2 == null
            || obj3 == null;
    }

    public IResult? SomeMethod(IDatabase? db, IAccount? account, IPermission perm)
    {
        if(FF.AnyNull(db, account, perm) return null;

        return db.GetInstance(perm.Token).GetPaymentsFor(account.Id);
        //       ^warning         ^warning                      ^warning
    }
```

Here, despite FF.AnyNull, the analyzer will provide null warnings for all the parameter. It doesn't know that `FF.AnyNull` already ensured things are fine.

**With annotations**

```csharp
    public static bool AnyNull([NotNullWhen(false)] object? obj1,
                               [NotNullWhen(false)] object? obj2,
                               [NotNullWhen(false)] object? obj3)
    {
        return obj1 == null
            || obj2 == null
            || obj3 == null;
    }

    public IResult? SomeMethod(IDatabase? db, IAccount? account, IPermission perm)
    {
        if(FF.AnyNull(db, account, perm) return null;

        return db.GetInstance(perm.Token).GetPaymentsFor(account.Id);
    }
```

Here, the annotations tell the analyzer: "Listen I know you cannot tell but ***trust me***, `FF.AnyNull` returned `false` and that means that I, the developer, am promising you that ***I*** can tell: `db`, `account` and `perm` are not null."

If the analyzer could understand the code, we wouldn't *need* the annotations, the analyzer would just do the right thing automatically.

And I would just like to not have to write 10 overloads for AnyNull when we already have `params` in the language. I want to write

```csharp
    public static bool AnyNull([ElementsNotNullWhen(false)] IReadOnlyCollection<object?>? objects)
    {
        if(objects == null) return true;
        foreach(var o in objects)
            if(o == null) return true;
        return false;
    }

    public static bool AnyNull([ElementsNotNullWhen(false)] params object?[]? objects)
    {
        return AnyNull( (IReadOnlyCollection<object?>?)objects);
    }
```

And have that work because the analyzer trusts the annotation. It won't have to understand the loop, it will just have to trust that I didn't lie in the annotation - which it *already has to trust anyway* because that's the point of the annotations.

**Reply by TahirAhmadov** [reply]
You're forgetting about this:
```cs
public static bool AnyNull([NotNullWhen(false)] object? obj1,
                               [NotNullWhen(false)] object? obj2,
                               [NotNullWhen(false)] object? obj3)
{
  return false; // warnings - obj1/2/3 may be null
}
```

**Reply by SRNissen** [reply]
I don't know what you're getting at - it seems to work as expected when I test it?

```csharp

    //In class FF
    public static bool AnyNull([NotNullWhen(false)] object? obj1,
                               [NotNullWhen(false)] object? obj2,
                               [NotNullWhen(false)] object? obj3)
    {
        return false;
    }

    //In test project
    [Test]
    public void AnyNull()
    {
        string? str = null;
        List<object>? list = null;
        object? obj = null;

            if (FF.AnyNull(str, list, obj)) return;
		
        list.Add(str); // throws System.NullReferenceException as expected, ending the test
        list.Add(obj);

        Assert.That(list, Has.Count.EqualTo(2));
    }
```

Despite compiling at /W4, this code has no warnings when using the AnyNull implementation you provided, exactly like `[NotNullWhen(false)]` advertises.

**Reply by TahirAhmadov** [reply]
Strange, it works differently when `ref` is specified on these parameters. Perhaps somebody who maintains the analyzer around `NotNullWhenAttribute` can pitch in?

**Reply by SRNissen** [reply]
"Huh."

Compiling example to work from: https://godbolt.org/z/5r354jbfq

```csharp
using System;
using System.Diagnostics.CodeAnalysis;

class Program
{
    public static bool AnyNull([NotNullWhen(false)] object? obj1,
                               [NotNullWhen(false)] object? obj2)
    {
        return false;
    }

    static void Main()
    {
        string? greeting = "Hello";
        string? subject = "world";

        if(!AnyNull(greeting, subject))
            Console.WriteLine(greeting.Length + subject.Length);

        greeting = null;
        subject = null;

        if(!AnyNull(greeting, subject))
            Console.WriteLine(greeting.Length + subject.Length);
    }
}
```

**Reply by TahirAhmadov** [reply]
Look at this:
https://sharplab.io/#v2:EYLgtghglgdgPgWAFAAEAMACFBGAdAESggHMYB7AZwBcoBjC3AYTIBMBTAQRggBsBPClAoBuZMhQBmLACYMjDAG9kGFVikoALBgCyACgCUi5aoC+xlZKzYAbBmBkyPDFz4A5AK48eugNquyVB5eAOoAFmwwulQATu5s+gC6GNFsAGYYZMAAVmy0VAD8GdnYADTmqhWVVdWqfgFBPGERUbHxSSnpmTl5hV3SZUg1Q8MVdYGejeGRMXGJyWlF3QWLEvrlSoPVUKm6XdgYALwHGDAThigA7BgzbKKbVdu72bJHJ2dYVzd3NY9dUq+nLznT6tb7VS4YVK8Ci3cpmJDIExAA=

**Reply by SRNissen** [reply]
That's... interesting.

I'm on dotnet 7, both with the code I'm writing that made me start this feature request, and on the Godbolt example, my compiler is (checks...) Roslyn, and I'm on debug mode with warning level 4 and warnings as errors.

Let me try changing some of those around, see if that makes a difference.

EDIT: No I can't seem to get it to warn on this. Keyword `ref` might make a real difference here?

EDIT 2: Alright - rather than getting non-`ref` to warn, instead I tried getting `ref` to *not* warn and that's doable. You go back a few minor versions and it stops warning. I think the lesson here is: If the analyzer can *tell* that your annotation is bogus, it will warn you. If your code is too complex for the analyzer to tell, it will trust you to use the annotations correctly.

If so, that would be fine by me - I *already* only need those annotations for situations where the analyzer cannot tell that the null check has been performed.

EDIT: **Dotnet** 7, not C# 7.
